lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
   pathHasContext = builtins.hasContext or (lib.hasPrefix builtins.storeDir);
 
   canCleanSource = src: src ? _isLibCleanSourceWith || !(pathHasContext (toString src));
+
+  # Splits a filesystem path into its components.
+  splitPath = path: lib.splitString "/" (toString path);
+
+  # Turns a list of path components into a tree, e.g.
+  #
+  #     ["a" "b" "c1"]
+  #     ["a" "b" "c2"]
+  #     ["a" "b" "c3"]
+  #     ["a" "x"     ]
+  #
+  # becomes:
+  #
+  #     { a = { b = { c1 = null; c2 = null; c3 = null }; x = null; }; }
+  pathComponentsToTree = paths: with lib;
+    foldl (tree: path: recursiveUpdate tree (setAttrByPath path null)) {} paths;
+
+  # Returns true if and only if any prefix of the given `path` leads to a leaf
+  # (`null`) in the given `tree` (nested attrset).
+  # That is: "If we go down the tree by the given path, do we hit a leaf?"
+  #
+  # Example: For the tree
+  #
+  #     a
+  #       b-leaf
+  #       c-leaf
+  #       d
+  #         e-leaf
+  #     f-leaf
+  #
+  # represented as attrset
+  #
+  #     { a = { b = null; c = null; d = { e = null; }; }; f = null; }
+  #
+  # we have
+  #
+  #     isPrefixOfLeafPath [a] == true
+  #     isPrefixOfLeafPath [x] == false
+  #     isPrefixOfLeafPath [a b] == true
+  #     isPrefixOfLeafPath [a b c] == true
+  #     isPrefixOfLeafPath [a b c x] == true
+  #     isPrefixOfLeafPath [a d] == false
+  isPrefixOfLeafPath = path: tree:
+    if tree == null
+      then true
+      else
+        if path == []
+          then false
+          else
+            let
+              component = builtins.head path;
+              restPath = builtins.tail path;
+            in
+              if !(builtins.hasAttr component tree)
+                then false
+                else
+                  let
+                    subtree = builtins.getAttr component tree;
+                  in
+                    isPrefixOfLeafPath restPath subtree;
+
+
+
+  # See `explicitFilterSource` for an example of this this filter.
+  #
+  # You can use this filter standalone (with `builtins.filterSource`
+  # or better, `builtins.path` with explicitly given `name`) when
+  # you want to combine it with other filters.
+  explicitSourceFilter =
+    {
+      # List of dirs under which all recursively contained files are taken in
+      # (unless a file is filtered by other arguments).
+      # Dirs that match explicitly are immediately taken in.
+      srcDirs ? [],
+      # Explicit list of files that should be taken in.
+      srcFiles ? [],
```

Comment:
Maybe the names `includeDirs` and `includeFiles` or similar would be a better fit.

Reply:
Done.